### PR TITLE
Engine v2

### PR DESCRIPTION
**Pull request by daniel-nichter**

### CODE REVIEW

**Review by Areson — COMMENTED** [review]
This generally looks good, though it's harder to verify all of the engine changes. I'm going to run another integration test on this and see how it goes.

**Review by joycse06 — APPROVED** [review]

File: event/event.go [hunk 1]
```diff
@@ -32,34 +32,57 @@ type Receiver interface {
...
+	return
 }
 
 // receiver is the private package Receiver that the public packages below use.
-// It defaults to a Log type receiver (set by Server.Start), but users can call
-// SetReceiver (before Server.Start) to override.
-var receiver Receiver = Log{}
+// It defaults to a Log type receiver (set in Server.Boot), but users can call
+// SetReceiver (before Server.Boot) to override.
+var receiver Receiver = Log{internal: true}
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply event/event.go]: Addressed by https://github.com/cashapp/blip/commit/8ba9aa4d59a0b7953e5ebf2eba449c294cc9e225

File: event/event.go [hunk 2]
```diff
@@ -32,34 +32,57 @@ type Receiver interface {
...
-
 // SetReceiver sets the receiver used by Blip to handle events. The default
 // receiver is Log. To override the default, call this function to set a new
 // receiver before calling Server.Boot.
-func SetReceiver(r Receiver) {
-	// Don't override becuase it's set only once. If user calls before
-	// Server.Start, then we keep their receiver. Else, Server.Start calls
-	// to set a built-in Log receiver.
-	once.Do(func() { receiver = r })
+func SetReceiver(r Receiver, override bool) {
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply event/event.go]: Addressed by https://github.com/cashapp/blip/commit/8ba9aa4d59a0b7953e5ebf2eba449c294cc9e225

File: event/event.go [hunk 3]
```diff
@@ -32,34 +32,57 @@ type Receiver interface {
...
+	// Is r the default Log receiver set below (outside this func)?
+	// Yes if r type = Log and that Log.internal = true. If Log.internal = false,
+	// then it's the Log set in Server.Boot because that pkg can't see/set the
+	// private Log.internal field.
+	defaultLog := false
+	if lr, ok := receiver.(Log); ok && lr.internal {
+		defaultLog = true
+	}
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply event/event.go:52]: Addressed by https://github.com/cashapp/blip/commit/8ba9aa4d59a0b7953e5ebf2eba449c294cc9e225

File: event/event.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply event/event.go]: Resolved the other related comments. Please resolve this one if new code and comments are 👍

**Review by Areson — APPROVED** [review]
Integration test looks good.

**Pull request merged**